src/casts/specify-representation.rkt

```diff
@@ -443,33 +416,46 @@ but a static single assignment is implicitly maintained.
     (define mk-crcn-label (Code-Label mk-crcn))
     (define mk-tuple-crcn-c : D0-Code
-      (code$ (t1 t2 l i count)
-        (If (op$ = i count)
-            (begin$
-              (assign$ crcn
-                (op$ Alloc (sr-plus count COERCION-TUPLE-ELEMENTS-OFFSET)))
-              (assign$ tagged-count
-                (op$ + (op$ %<< count COERCION-SECOND-TAG-SHIFT)
-                     COERCION-TUPLE-SECOND-TAG))
-              (sr-array-set! crcn COERCION-TUPLE-COUNT-INDEX tagged-count)
-              (sr-tag-value crcn COERCION-MEDIATING-TAG))
-            (begin$
-              (assign$ t1a
-                (sr-tagged-array-ref
-                 t1 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
-              (assign$ t2a
-                (sr-tagged-array-ref
-                 t2 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
-              (assign$ val-crcn (App-Code mk-crcn-label `(,t1a ,t2a ,l)))
-              (assign$ tmp-crcn
-                (App-Code
-                 mk-tuple-crcn-label `(,t1 ,t2 ,l ,(sr-plus (Quote 1) i) ,count)))
-              (sr-tagged-array-set!
-               tmp-crcn
-               COERCION-MEDIATING-TAG
-               (sr-plus COERCION-TUPLE-ELEMENTS-OFFSET i)
-               val-crcn)
-              tmp-crcn))))
+      (code$ (t1 t2 l)
+        (assign$ t1-count
+          (sr-tagged-array-ref t1 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))
+        (assign$ t2-count
+          (sr-tagged-array-ref t2 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))
+        (assign$ smaller-count (If (op$ < t1-count t2-count) t1-count t2-count))
+        (assign$ bigger-count (If (op$ > t1-count t2-count) t1-count t2-count))
+        (assign$ crcn
+          (op$ Alloc (op$ + bigger-count COERCION-TUPLE-ELEMENTS-OFFSET)))
+        (assign$ tagged-count
+          (op$ + (op$ %<< bigger-count COERCION-SECOND-TAG-SHIFT)
+               COERCION-TUPLE-SECOND-TAG))
+        (assign$ iters (op$ + TYPE-TUPLE-ELEMENTS-OFFSET smaller-count))
+        (repeat$ (i TYPE-TUPLE-ELEMENTS-OFFSET iters) (_ UNIT-IMDT)
+          (assign$ t1a (sr-tagged-array-ref t1 TYPE-TUPLE-TAG i))
+          (assign$ t2a (sr-tagged-array-ref t2 TYPE-TUPLE-TAG i))
+          (assign$ val-crcn (app-code$ mk-crcn-label t1a t2a l))
+          (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                            COERCION-TUPLE-ELEMENTS-OFFSET))
+          (sr-array-set! crcn c-i val-crcn))
+        (cond$
+         [(op$ > t1-count t2-count)
+          (assign$ iters (op$ + t1-count TYPE-TUPLE-ELEMENTS-OFFSET))
+          (assign$ i-init (op$ + t2-count TYPE-TUPLE-ELEMENTS-OFFSET))
+          (repeat$ (i i-init iters) (_ UNIT-IMDT)
+            (assign$ t1a (sr-tagged-array-ref t1 TYPE-TUPLE-TAG i))
+            (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                              COERCION-TUPLE-ELEMENTS-OFFSET))
+            (sr-array-set! crcn c-i COERCION-IDENTITY-IMDT))]
```

Comment:
Here we are invoking the rules of subtyping and we should probably document this or something. I believe the first branch (> t1 t2) is okay because t2 is a subtype of t1. We may want to not create the ID coercions, for optimization and heap representation purposes. This is dependent on how tuple coercion application is implemented so we should check and document this in both places.

Reply:
make_coercion will create an ID coercion if the two types are equal, we have hashconsing now. Otherwise, done.

Reply:
On a second thought, composing a tuple with a subtype tuple can still result in unnecessarily allocated id crcn.

Reply:
Done.